src/wkt/src/internal/enums.rs

```diff
@@ -0,0 +1,232 @@
+    pub fn value(&self) -> Option<i32> {
+        match self {
+            Self::Integer(x) => Some(*x),
+            Self::String(_) => None,
+        }
+    }
+    pub fn name(&self) -> Option<&str> {
+        match self {
+            Self::Integer(_) => None,
+            Self::String(x) => Some(x.as_str()),
+        }
+    }
+}
+
+impl serde::ser::Serialize for UnknownEnumValue {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            Self::Integer(x) => serializer.serialize_i32(*x),
+            Self::String(x) => serializer.serialize_str(x.as_str()),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct EnumVisitor<'lf, T> {
+    name: &'lf str,
+    _unused: PhantomData<T>,
+}
+
+impl<'lf, T> EnumVisitor<'lf, T> {
+    pub fn new(name: &'lf str) -> Self {
+        Self {
+            name,
+            _unused: Default::default(),
+        }
+    }
+}
+
+impl<T> serde::de::Visitor<'_> for EnumVisitor<'_, T>
+where
+    T: From<i32> + for<'a> From<&'a str>,
+{
+    type Value = T;
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(T::from(value))
+    }
+    fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        if value > (i32::MAX as u64) {
+            return Err(E::custom(format!(
+                "out of range enum value {value} for {}",
+                self.name,
+            )));
+        }
+        Ok(T::from(value as i32))
+    }
+    fn visit_i64<E>(self, value: i64) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        if value < (i32::MIN as i64) || value > (i32::MAX as i64) {
+            return Err(E::custom(format!(
+                "out of range enum value {value} for {}",
+                self.name,
+            )));
+        }
+        Ok(T::from(value as i32))
+    }
+    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+        formatter.write_str(&format!(
+            "a {} enum value in string or integer format",
+            self.name
+        ))
+    }
+}
+
+pub fn display_enum(
+    f: &mut std::fmt::Formatter<'_>,
+    name: Option<&str>,
+    value: Option<i32>,
+) -> Result<(), std::fmt::Error> {
+    match (name, value) {
+        (Some(n), Some(_)) => f.write_str(n),
+        (Some(n), None) => f.write_str(n),
+        (None, Some(v)) => write!(f, "{}", v),
+        (None, None) => unreachable!("enums must have a numeric or string value"),
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use serde_json::json;
+    use test_case::test_case;
+
+    #[test]
+    fn unknown_enum_value_accessors() {
+        let u = UnknownEnumValue::Integer(123);
+        assert_eq!(u.value(), Some(123));
+        assert_eq!(u.name(), None);
+
+        let u = UnknownEnumValue::String("123".into());
```

Comment:
nit: can we call this "foo" or something? Just so the reader doesn't think it is related to `123`, the value.

Here and in the next test.

Reply:
Done.